### DISCUSSION TITLE
Eventarc API Firestore Triggers in european region not working?

### DISCUSSION THREAD

**Discussion by blaueeiner** [GoogleCloudPlatform/functions-framework-dart#249] [Q&A] [answered]
### What I want to achieve
I am trying to setup an Eventarc trigger. A specific Firestore action (deletion of a document) should trigger a Cloud Run service instance.

### My problem
The service should be provided in europe but...

**Eventarc** is only supported in the following regions:

europe-north1
europe-west1
europe-west4
(https://cloud.google.com/eventarc/docs/locations#europe)

**Firestore** is only supported in the following regions:

europe-west2
europe-west3
europe-west6
europe-central12
(https://firebase.google.com/docs/firestore/locations)

When I tried setting up the trigger on the Cloud Run Service, it prompted me to select a region from where (i.e. the region) the trigger should originate and receive events from. My Firestore region is set to europe-west3. Since Eventarc is not supported in this region I was trying to find our if I can get Firestore into a supported Eventarc region in Europe but there is literally none. Does that theoretically mean that Eventarc does not support Firestore in Europe at all?

There is the option to select all regions but I assume that this just means receiving events from all the supported regions, right?

![](https://i.stack.imgur.com/uuXRc.png)

Is there any way I can get this working or any official plans from Google about when to support this?

**Comment by nickmeinhold** [comment]
Hi @blaueeiner!  

It is my ([unconfirmed](https://github.com/googleapis/google-cloudevents/issues/184#issuecomment-852740108)) understanding that, regardless of region, deletion of a document is not currently one of the [Events supported by Eventarc](https://cloud.google.com/eventarc/docs/reference/supported-events#cloud-firestore).

I know that doesn't answer your query in regards to the zones but I thought it may be useful info.  I hope we get Firestore document events via Eventarc soon!

**Reply by blaueeiner** [reply]
Thanks for your reply. I assumed it is supported (at least partially) since I have been able to select the event "google.longrunning.Operations.DeleteOperation" from the category "Cloud Firestore" while trying to create the trigger. Just the region selection blocked me from continuing. Or am I wrong and "google.longrunning.Operations.DeleteOperation" is not what I want for the trigger on document deletion? :)

**Reply by nickmeinhold** [reply]
Yeah I think that's a very reasonable conclusion.  From that issue I linked to, where the guy was trying to get auth triggers working, he eventually got events coming through but they didn't have any user data in them and my takeaway from the issue was that if an event isn't on that eventarc list it possibly/definitely won't work end to end... it's all conjecture though so I'll stop adding noise to your post 🙂 I did do some experimenting at one point that from memory backed up the idea that it's not supported yet but I hope you're right.  If I find some time to do some more experimenting I'll come back and post what I find, but hopefully you get a definitive answer in the meantime.  Good luck with it 👍

**Reply by blaueeiner** [reply]
I will keep everybody in this discussion up to date on further progress and research :)

**Reply by blaueeiner** [reply]
For now I will create lightweight Firestore Triggered Cloud Functions written in Typescript which will simply forward the data of the Firestore trigger via Cloud Pub/Sub. On the Cloud Run Service I will have a Cloud Pub/Sub trigger configured.

**Reply by nickmeinhold** [reply]
Yeah that's what I've done for now - I've only done Firebase Auth so far, would love to see it if you're happy to share when it's done.  There's a relevant [issue here](https://github.com/GoogleCloudPlatform/functions-framework-dart/issues/239) in case that's helpful.

**Reply by nickmeinhold** [reply]
@blaueeiner someone pointed out to me that [audit logs for document events](https://cloud.google.com/firestore/docs/audit-logging#upcoming-audit-logs) will be available on June 30th.

**Reply by blaueeiner** [reply]
@nickmeinhold Thank you very much for the information. I will try it out and share my experience using it 💪

**Reply by nickmeinhold** [reply]
Hi again, just passing on - the same person pointed out that the docs have now changed to say: "Audit logs for the following operations will become available soon."  🤷‍♂️

**Reply by and3rson** [reply]
Any information on when will EventArc natively support Firestore? It's been several years, and writing boilerplate Cloud Functions that proxy the data to PubSub or Cloud Run is a pain, especially with local emulator.

**Reply by kevmoo** [reply]
I'm asking around a bit internally...

**Reply by and3rson** [reply]
@kevmoo Thanks! Any news by any chance?..

**Comment by jimmyff** [comment, marked as answer]
It looks like you can now create Event Arc triggers for:

- google.firestore.v1.Firestore.CreateDocument
- google.firestore.v1.Firestore.UpdateDocument
- google.firestore.v1.Firestore.DeleteDocument
- google.firestore.v1.Firestore.Write

See:
![image](https://user-images.githubusercontent.com/298850/128351602-0fa3508a-be62-46b9-844d-c5c676a0a38c.png)

Problem is the triggers don't see to be working for me, I'm trying to get the CreateDocument one working. Ideally I'd like to limit it to my `account` collection and I'm unclear what Resource Name that would require. Right now I'm just trying to get it trigger for anything.

I've [posted on stackoverflow](https://stackoverflow.com/questions/68654256/gcp-eventarc-trigger-for-firestore-doc-creation-limited-to-specific-collection) to try and get an answer. If anyone has any luck please share

**Reply by nickmeinhold** [reply]
fwiw data access still seems to be under "[Upcoming audit logs](https://cloud.google.com/firestore/docs/audit-logging#upcoming-audit-logs)" for now

**Reply by jimmyff** [reply]
Ah, maybe they've put the endpoints live a little early. Looks like they must be just around the corner at least.

**Reply by jimmyff** [reply]
Looks like the docs are now listing them as a '**preview**' also I noticed there a load of new endpoints under the namespace: `google.firestore.v1beta1.$operation`.

I've not tested them yet but will have play soon

**Reply by jimmyff** [reply]
So the only endpoint I'm having any success with is: `google.firestore.v1.Firestore.Write` I have an event arc trigger firing up my Cloud Run instance successfully. Insert, update, delete all seem to trigger this Write event. I am however unclear on how I should filter this to a specific collection.

When creating the trigger you can specify a `Resource`, however the resource for all my `Write` events are the same: `projects/$appName/databases/(default)`.

The `resource` field in the json payload looks like the following:
```
resource: {
  labels: {
    method: "google.firestore.v1.Firestore.Write",
    project_id: "$appName",
    service: "firestore.googleapis.com"
  },
  type: "audited_resource"
}
```

Feels like this is so close, we can almost replicate the functionality of Firestore Functions. If anyone figures out how to filter this to specific collections, and possibly different types of writes (insert, delete etc) please share details!

**Comment by kevmoo** [comment]
Please see https://cloud.google.com/firestore/docs/eventarc !!!
